File: ad-server/app/services/event_geo.py
"""Geographic filtering for station events (mirrors ad campaign rules)."""
from __future__ import annotations

from typing import Optional

from app.schemas.events import StationEvent


def filter_events_for_country(
    events: list[StationEvent],
    country: Optional[str],
) -> list[StationEvent]:
    """
    - Event with no country_code: worldwide (any detected country).
    - Event with country_code: only listeners in that ISO country.
    - Unknown listener country: only worldwide events (no country_code).
    """
    if not country:
        return [e for e in events if not _event_country(e)]

    code = country.strip().upper()
    return [
        e
        for e in events
        if not _event_country(e) or _event_country(e) == code
    ]


def _event_country(event: StationEvent) -> Optional[str]:
    if not event.country_code:
        return None
    return event.country_code.strip().upper() or None

File: app/src/utils/eventGeo.ts
import type { StationEvent } from '../types';

/** Same rules as ad-server event_geo.filter_events_for_country */
export function eventVisibleForCountry(
  event: StationEvent,
  country: string | null | undefined,
): boolean {
  const code = event.countryCode?.trim().toUpperCase();
  if (!code) return true;
  if (!country?.trim()) return false;
  return code === country.trim().toUpperCase();
}

export function filterEventsForCountry(
  events: StationEvent[],
  country: string | null | undefined,
): StationEvent[] {
  return events.filter((e) => eventVisibleForCountry(e, country));
}
